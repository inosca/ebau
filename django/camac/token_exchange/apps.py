from django.apps import AppConfig


class TokenExchangeConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "camac.token_exchange"
