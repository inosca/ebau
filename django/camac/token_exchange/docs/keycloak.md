# Keycloak configuration

## 1. Create new client "token-exchange"

![Create client](./images/create-client.png)

## 2. Grant service account role "manage-users" to "token-exchange" client

![Grant role](./images/grant-role.png)

## 3. Create new positive client policy "token-exchange"

![Create policy step 1](./images/create-policy-1.png)
![Create policy step 2](./images/create-policy-2.png)

## 4. Grant permissions for token exchange on "portal" client and assign policy "token-exchange"

![Grant token exchange permission step 1](./images/grant-token-exchange-1.png)
![Grant token exchange permission step 2](./images/grant-token-exchange-2.png)

## 5. Grant permissions for impersonate on user permission tab and assign policy "token-exchange"

![Grant impersonate permission step 1](./images/grant-impersonate-1.png)
![Grant impersonate permission step 2](./images/grant-impersonate-2.png)

## 6. Create new user attribute "LoT"

![Create user attribute step 1](./images/lot-attribute-1.png)
![Create user attribute step 2](./images/lot-attribute-2.png)

## 7. Create new client scope "token-exchange"

![Create client scope step 1](./images/create-client-scope-1.png)
![Create client scope step 2](./images/create-client-scope-2.png)

## 8. Add new token mapper "LoT" to client scope "token-exchange"

![Add LoT mapper step 1](./images/lot-mapper-1.png)
![Add LoT mapper step 2](./images/lot-mapper-2.png)

## 9. Assign client scope "token-exchange" to client "portal"

![Assign client scope step 1](./images/assign-client-scope-1.png)
![Assign client scope step 2](./images/assign-client-scope-2.png)
