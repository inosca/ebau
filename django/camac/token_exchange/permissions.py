from django.conf import settings
from jwt import decode
from rest_framework.authentication import get_authorization_header
from rest_framework.permissions import BasePermission


def get_lot(request):
    try:
        _, token = get_authorization_header(request).split()
        token_data = decode(token, options={"verify_signature": False})

        return int(token_data["lot"])
    except (ValueError, KeyError):
        return None


def has_required_lot(request):
    if not settings.ENABLE_TOKEN_EXCHANGE:
        return True

    from camac.user.permissions import get_role_name

    lot = get_lot(request)
    role = get_role_name(request.group)

    if role not in settings.TOKEN_EXCHANGE_LOT_MAPPING:
        return True

    if lot is None:
        return False

    return lot >= settings.TOKEN_EXCHANGE_LOT_MAPPING[role]


class RequireLoT(BasePermission):
    def has_permission(self, request, view):
        return has_required_lot(request)
