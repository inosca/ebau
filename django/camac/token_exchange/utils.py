from json import JSONDecodeError

from django.conf import settings
from django.utils import timezone
from jwcrypto.common import JWException, base64url_decode, json_decode
from jwcrypto.jwk import JWK
from jwcrypto.jwt import JWT


def extract_headers(raw_token: str) -> dict:
    try:
        return json_decode(base64url_decode(raw_token.split(".")[0]))
    except (IndexError, ValueError, JSONDecodeError):  # pragma: no cover
        raise JWException("Could not parse headers")


def verify_jwe_headers(headers: dict) -> None:
    if headers.get("alg") != "dir":
        raise JWException("Invalid JWE alg header")
    elif headers.get("enc") not in ["A256CBC-HS512", "A128CBC-HS256"]:
        raise JWException("Invalid JWE enc header")
    elif headers.get("cty") != "JWT":
        raise JWException("Invalid JWE cty header")
    elif headers.get("typ") != "JWT":
        raise JWException("Invalid JWE typ header")


def verify_jwt_headers(headers: dict) -> None:
    if headers.get("alg") != "HS256":
        raise JWException("Invalid JWT alg header")
    elif headers.get("typ") != "JWT":
        raise JWException("Invalid JWT typ header")


def verify_jwt_claims(claims: dict) -> None:
    max_validity_period = settings.TOKEN_EXCHANGE_MAX_VALIDITY_PERIOD

    if claims["iat"] > timezone.now().timestamp():
        raise JWException("Issued at is in the future (iat)")
    elif max_validity_period and claims["exp"] - claims["nbf"] > max_validity_period:
        raise JWException(
            f"Validity period (exp - nbf) of token is longer than {max_validity_period} seconds"
        )


def extract_jwt_data(encrypted_token: str) -> dict:
    verify_jwe_headers(extract_headers(encrypted_token))

    signed_token = JWT(
        jwt=encrypted_token,
        key=JWK.from_password(settings.TOKEN_EXCHANGE_JWE_SECRET),
        expected_type="JWE",
    )

    verify_jwt_headers(extract_headers(signed_token.claims))

    token = JWT(
        jwt=signed_token.claims,
        key=JWK.from_password(settings.TOKEN_EXCHANGE_JWT_SECRET),
        expected_type="JWS",
        check_claims={
            "exp": None,
            "nbf": None,
            "iat": None,
            "iss": settings.TOKEN_EXCHANGE_JWT_ISSUER,
        },
    )

    verify_jwt_claims(json_decode(token.claims))

    return json_decode(token.claims)


def extract_sync_data(jwt_data: dict) -> dict:
    data = {
        keycloak_property: jwt_data.get(jwt_property)
        for jwt_property, keycloak_property in settings.TOKEN_EXCHANGE_JWT_SYNC_PROPERTIES.items()
    }

    organisationName = data.pop("organisationName")

    if organisationName:
        data["firstName"] = ""
        data["lastName"] = organisationName

    data["attributes"] = {"lot": data.pop("lot")}

    return data


def build_username(jwt_data: dict):
    return settings.TOKEN_EXCHANGE_USERNAME_PREFIX + str(
        jwt_data[settings.TOKEN_EXCHANGE_JWT_IDENTIFIER_PROPERTY]
    )
