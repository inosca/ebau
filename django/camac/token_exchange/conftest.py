import pytest
from django.utils import timezone
from faker import Faker
from jwcrypto.jwk import JWK
from jwcrypto.jwt import JWT
from rest_framework.test import APIClient

from camac.conftest import reload_urlconf

fake = Faker()


def encode_token(data, jwt_secret, jwe_secret):
    signed_token = JWT(header={"alg": "HS256", "typ": "JWT"}, claims=data)
    signed_token.make_signed_token(JWK.from_password(jwt_secret))

    encrypted_token = JWT(
        header={"alg": "dir", "enc": "A256CBC-HS512", "cty": "JWT", "typ": "JWT"},
        claims=signed_token.serialize(),
    )
    encrypted_token.make_encrypted_token(JWK.from_password(jwe_secret))

    return encrypted_token.serialize()


@pytest.fixture(autouse=True)
def enable(settings):
    settings.ENABLE_TOKEN_EXCHANGE = True
    settings.KEYCLOAK_URL = "http://ebau-keycloak.local/auth"
    settings.KEYCLOAK_OIDC_TOKEN_URL = (
        "http://ebau-keycloak.local/auth/realms/ebau/protocol/openid-connect/token"
    )
    settings.TOKEN_EXCHANGE_MAX_VALIDITY_PERIOD = 60

    reload_urlconf("camac.urls")
    reload_urlconf("camac.token_exchange.urls")


@pytest.fixture
def jwt_client():
    return APIClient()


@pytest.fixture
def jwt_token_data(settings):
    now = timezone.now().timestamp()

    return {
        "profileId": 1,
        "firstName": "John",
        "name": "Doe",
        "email": "john.doe@acme.com",
        "exp": now + settings.TOKEN_EXCHANGE_MAX_VALIDITY_PERIOD - 1,
        "iss": settings.TOKEN_EXCHANGE_JWT_ISSUER,
        "iat": now - 1,
        "nbf": now,
        "lot": "1",
    }


@pytest.fixture
def jwt_token(jwt_token_data, settings):
    return encode_token(
        jwt_token_data,
        settings.TOKEN_EXCHANGE_JWT_SECRET,
        settings.TOKEN_EXCHANGE_JWE_SECRET,
    )


@pytest.fixture
def invalid_jwt_token(jwt_token_data, settings):
    return encode_token(
        jwt_token_data,
        settings.TOKEN_EXCHANGE_JWT_SECRET,
        "abcdefghijklmnopqrstuvwabcdefghijklmnopqrstuvwabcdefghijklmnopqr",
    )
