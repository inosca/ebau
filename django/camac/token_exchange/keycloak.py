from typing import Union

import requests
from django.conf import settings

from camac.token_exchange.utils import extract_sync_data
from camac.utils import build_url


class KeycloakClient:
    def __init__(self):
        self.token = self.get_token()

    def get_token(self) -> str:
        response = requests.post(
            settings.KEYCLOAK_OIDC_TOKEN_URL,
            {
                "grant_type": "client_credentials",
                "client_id": settings.TOKEN_EXCHANGE_CLIENT,
                "client_secret": settings.TOKEN_EXCHANGE_CLIENT_SECRET,
            },
        )

        response.raise_for_status()

        return response.json()["access_token"]

    def get_user(self, username: str) -> Union[str, None]:
        response = requests.get(
            build_url(
                settings.KEYCLOAK_URL,
                "admin/realms",
                settings.KEYCLOAK_REALM,
                f"users?username={username}",
            ),
            headers={"authorization": f"Bearer {self.token}"},
        )

        response.raise_for_status()

        result = response.json()

        return result[0] if len(result) > 0 else None

    def create_user(self, username: str, data: dict) -> None:
        response = requests.post(
            build_url(
                settings.KEYCLOAK_URL,
                "admin/realms",
                settings.KEYCLOAK_REALM,
                "users",
            ),
            json={
                "username": username,
                "enabled": True,
                **extract_sync_data(data),
            },
            headers={
                "authorization": f"Bearer {self.token}",
                "content-type": "application/json",
            },
        )

        response.raise_for_status()

    def update_user(self, user_id: str, data: dict) -> None:
        response = requests.put(
            build_url(
                settings.KEYCLOAK_URL,
                "admin/realms",
                settings.KEYCLOAK_REALM,
                "users",
                user_id,
            ),
            json=extract_sync_data(data),
            headers={
                "authorization": f"Bearer {self.token}",
                "content-type": "application/json",
            },
        )

        response.raise_for_status()

    def update_or_create_user(self, username: str, data: dict) -> bool:
        user = self.get_user(username)

        if user:
            self.update_user(user["id"], data)
        else:
            self.create_user(username, data)

        return bool(user)

    def token_exchange(self, username: str) -> dict:
        response = requests.post(
            settings.KEYCLOAK_OIDC_TOKEN_URL,
            {
                "grant_type": "urn:ietf:params:oauth:grant-type:token-exchange",
                "client_id": settings.TOKEN_EXCHANGE_CLIENT,
                "client_secret": settings.TOKEN_EXCHANGE_CLIENT_SECRET,
                # https://github.com/keycloak/keycloak/issues/17668
                # "audience": settings.KEYCLOAK_PORTAL_CLIENT,
                "requested_token_type": "urn:ietf:params:oauth:token-type:refresh_token",
                "requested_subject": username,
                "scope": "openid",
            },
        )

        response.raise_for_status()

        return response.json()
