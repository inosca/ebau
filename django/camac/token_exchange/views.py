import json
import logging

from jwcrypto.common import JWException
from requests import HTTPError, JSONDecodeError
from rest_framework import exceptions, response, status
from rest_framework.parsers import JSONParser
from rest_framework.renderers import JSONRenderer
from rest_framework.views import APIView

from camac.token_exchange.keycloak import KeycloakClient
from camac.token_exchange.utils import build_username, extract_jwt_data

logger = logging.getLogger(__name__)


class TokenExchangeView(APIView):
    parser_classes = [JSONParser]
    renderer_classes = [JSONRenderer]
    authentication_classes = []
    permission_classes = []

    def post(self, request, *args, **kwargs):
        jwt_token = request.data.get("jwt-token")

        if not jwt_token:
            raise exceptions.AuthenticationFailed("JWT token must be passed")

        try:
            jwt_data = extract_jwt_data(jwt_token)
            username = build_username(jwt_data)

            keycloak_client = KeycloakClient()
            keycloak_client.update_or_create_user(username, jwt_data)

            token = keycloak_client.token_exchange(username)
        except HTTPError as e:
            try:
                result = e.response.json()
                message = result.get("errorMessage", json.dumps(result))
            except JSONDecodeError:  # pragma: no cover
                message = e.response.text
            logger.error(message)
            raise exceptions.AuthenticationFailed("Token could not be exchanged")
        except JWException as e:
            logger.error(e)
            raise exceptions.AuthenticationFailed("Invalid token")

        return response.Response(
            token,
            status=status.HTTP_200_OK,
        )
