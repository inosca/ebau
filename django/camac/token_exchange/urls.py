from django.urls import re_path

from camac.token_exchange.views import TokenExchangeView

urlpatterns = [
    re_path(
        r"^auth/token-exchange", TokenExchangeView.as_view(), name="token-exchange"
    ),
]
