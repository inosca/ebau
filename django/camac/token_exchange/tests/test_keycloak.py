import jwt
import pytest

from camac.token_exchange.keycloak import KeycloakClient


@pytest.mark.vcr()
def test_keycloak_client(snapshot):
    client = KeycloakClient()

    username = "egov:123"
    user_data = {
        "firstName": "Test",
        "name": "User",
        "email": "email@example.com",
        "organisationName": "",
        "lot": "1",
    }

    # first call should create a new user
    assert not client.update_or_create_user(username, user_data)
    assert client.get_user(username) == snapshot

    # second call should update the existing user
    user_data["organisationName"] = "ACME Inc."

    assert client.update_or_create_user(username, user_data)
    assert client.get_user(username) == snapshot

    token = client.token_exchange(username)
    assert token == snapshot

    decoded_token = jwt.decode(
        token["access_token"], options={"verify_signature": False}
    )
    assert decoded_token == snapshot
