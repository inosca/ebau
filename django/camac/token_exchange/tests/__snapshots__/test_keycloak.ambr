# serializer version: 1
# name: test_keycloak_client
  dict({
    'access': dict({
      'impersonate': True,
      'manage': True,
      'manageGroupMembership': True,
      'mapRoles': True,
      'view': True,
    }),
    'attributes': dict({
      'lot': list([
        '1',
      ]),
    }),
    'createdTimestamp': 1726139446128,
    'disableableCredentialTypes': list([
    ]),
    'email': 'email@example.com',
    'emailVerified': True,
    'enabled': True,
    'firstName': 'Test',
    'id': 'fb36bee2-e06c-436d-ad40-bddcadacdf0c',
    'lastName': 'User',
    'notBefore': 0,
    'requiredActions': list([
    ]),
    'totp': False,
    'username': 'egov:123',
  })
# ---
# name: test_keycloak_client.1
  dict({
    'access': dict({
      'impersonate': True,
      'manage': True,
      'manageGroupMembership': True,
      'mapRoles': True,
      'view': True,
    }),
    'attributes': dict({
      'lot': list([
        '1',
      ]),
    }),
    'createdTimestamp': 1726139446128,
    'disableableCredentialTypes': list([
    ]),
    'email': 'email@example.com',
    'emailVerified': True,
    'enabled': True,
    'id': 'fb36bee2-e06c-436d-ad40-bddcadacdf0c',
    'lastName': 'ACME Inc.',
    'notBefore': 0,
    'requiredActions': list([
    ]),
    'totp': False,
    'username': 'egov:123',
  })
# ---
# name: test_keycloak_client.2
  dict({
    'access_token': 'eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI3WHJ1di0wLWo2RUdRNlpRTS13OHpDekFYTC1TSExLQ0RlLWExc1VKMGJNIn0.eyJleHAiOjE3MjYxNDEyNDYsImlhdCI6MTcyNjEzOTQ0NiwianRpIjoiNDU4YmJhMTItNDhhMy00ZDRiLWJlMjMtMTZjYThlNzQyYWJlIiwiaXNzIjoiaHR0cDovL2ViYXUta2V5Y2xvYWsubG9jYWwvYXV0aC9yZWFsbXMvZWJhdSIsImF1ZCI6WyJwb3J0YWwiLCJhY2NvdW50Il0sInN1YiI6ImVnb3Y6MTIzIiwidHlwIjoiQmVhcmVyIiwiYXpwIjoidG9rZW4tZXhjaGFuZ2UiLCJzaWQiOiJlMGI3YzQwNi1kNjBmLTQxMWItYjFiMy0zYTA5NmE0YzVlZjIiLCJhY3IiOiIxIiwiYWxsb3dlZC1vcmlnaW5zIjpbImh0dHA6Ly9lZ292LmxvY2FsIiwiaHR0cDovL2xvY2FsaG9zdDo0MjAwIiwiaHR0cDovL2ViYXUtcG9ydGFsLmxvY2FsIl0sInJlYWxtX2FjY2VzcyI6eyJyb2xlcyI6WyJvZmZsaW5lX2FjY2VzcyIsImRlZmF1bHQtcm9sZXMtZWJhdSIsInVtYV9hdXRob3JpemF0aW9uIl19LCJyZXNvdXJjZV9hY2Nlc3MiOnsiYWNjb3VudCI6eyJyb2xlcyI6WyJtYW5hZ2UtYWNjb3VudCIsIm1hbmFnZS1hY2NvdW50LWxpbmtzIiwidmlldy1wcm9maWxlIl19fSwic2NvcGUiOiJvcGVuaWQgZW1haWwgcHJvZmlsZSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJuYW1lIjoiQUNNRSBJbmMuIiwicHJlZmVycmVkX3VzZXJuYW1lIjoiZWdvdjoxMjMiLCJmYW1pbHlfbmFtZSI6IkFDTUUgSW5jLiIsImVtYWlsIjoiZW1haWxAZXhhbXBsZS5jb20ifQ.Hph2EFp-UbdAtLjQiunW3qaRNUtBCtpkomtyo4-qR5TB7w-Axy_k1j4geSd1FjnjmRuGzNJoxdPl9kYjVb5f5hcKg-H0lvQwJJSqvWNZGe9bSQm-c4Ykzyn36trOGxGVzLIL2UEMLjguJQYKUqdhttWK4cfEO--OSdvYA6s8UjhVbTRUGv65gXFxAV1qbT4Xrqxi95DK07P7fxXwhI9g1O5lwvJKORjSDZIo8chvrgzwYjFs1leR8jyh5YNheXQqQ-siZ2BONDptYctiq6OjwH4nvOL5WXGuEHQeMztV6HAGjDbj4Fiz5Uk20p2NA3QkCphsH1mVKYqXq-fy2e_d_g',
    'expires_in': 1800,
    'id_token': 'eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI3WHJ1di0wLWo2RUdRNlpRTS13OHpDekFYTC1TSExLQ0RlLWExc1VKMGJNIn0.eyJleHAiOjE3MjYxNDEyNDYsImlhdCI6MTcyNjEzOTQ0NiwianRpIjoiZmVlNjJmZDUtYjQzOS00OWFmLWE0ZjQtZGYxNWUxN2Q1MzAzIiwiaXNzIjoiaHR0cDovL2ViYXUta2V5Y2xvYWsubG9jYWwvYXV0aC9yZWFsbXMvZWJhdSIsImF1ZCI6InRva2VuLWV4Y2hhbmdlIiwic3ViIjoiZWdvdjoxMjMiLCJ0eXAiOiJJRCIsImF6cCI6InRva2VuLWV4Y2hhbmdlIiwic2lkIjoiZTBiN2M0MDYtZDYwZi00MTFiLWIxYjMtM2EwOTZhNGM1ZWYyIiwiYXRfaGFzaCI6Ik5kMTdfTzZhQkRwUWdrNGpqel9iUlEiLCJhY3IiOiIxIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsIm5hbWUiOiJBQ01FIEluYy4iLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJlZ292OjEyMyIsImZhbWlseV9uYW1lIjoiQUNNRSBJbmMuIiwiZW1haWwiOiJlbWFpbEBleGFtcGxlLmNvbSJ9.dwmKloeuGilpBZ6lROVJ_3tupVapDh8vr7K6Kx2BCRYG54lE-tOoSvuprkC1gUu5rBkMZuQQLY0zP1QTXaEGiAKBavsoK6SVWCx4BbcsHJr76Yf4BRanJz5knaehetREf5wbUc-Za8r5IWolJwVuUfMG7gvcCMb00DGvWqTqtyIYF_x_tKf-pcV-lVD_co4xt1RwFI0WBEshZhytRycADmdH8Fb6qWXygwNhgsJ33MtpLsntXa1BQkpH8FFtz2EDHl5BNDzgu80D3v4m9a89u2F9VPxae1_sjjW5NZpYKURFdgJVnzFnFnaQS5O0tc8NgZTyNjx1jK1stS0Dmqjnrg',
    'not-before-policy': 0,
    'refresh_expires_in': 18000,
    'refresh_token': 'eyJhbGciOiJIUzUxMiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJlNDJkOTc3NS01NjE3LTQ4ZTUtODFlOC1jMjk3NDEwNDg4ODkifQ.eyJleHAiOjE3MjYxNTc0NDYsImlhdCI6MTcyNjEzOTQ0NiwianRpIjoiMzc2ZmVlZWItNTczYS00OTE3LTg3NzMtYjQwYmRiYTgwMTY5IiwiaXNzIjoiaHR0cDovL2ViYXUta2V5Y2xvYWsubG9jYWwvYXV0aC9yZWFsbXMvZWJhdSIsImF1ZCI6Imh0dHA6Ly9lYmF1LWtleWNsb2FrLmxvY2FsL2F1dGgvcmVhbG1zL2ViYXUiLCJzdWIiOiJlZ292OjEyMyIsInR5cCI6IlJlZnJlc2giLCJhenAiOiJ0b2tlbi1leGNoYW5nZSIsInNpZCI6ImUwYjdjNDA2LWQ2MGYtNDExYi1iMWIzLTNhMDk2YTRjNWVmMiIsInNjb3BlIjoib3BlbmlkIHJvbGVzIGVtYWlsIGFjciB3ZWItb3JpZ2lucyBwcm9maWxlIGJhc2ljIn0.Rlj1nBXibBS7w2O1J7k8-mB4ObFhL5T-OwC1svGqjJu-aqRH32jxqCJF1lMv7xIKuqmtkpGrkL2u5oCpBd1i2w',
    'scope': 'openid email profile',
    'session_state': 'e0b7c406-d60f-411b-b1b3-3a096a4c5ef2',
    'token_type': 'Bearer',
  })
# ---
# name: test_keycloak_client.3
  dict({
    'acr': '1',
    'allowed-origins': list([
      'http://egov.local',
      'http://localhost:4200',
      'http://ebau-portal.local',
    ]),
    'aud': list([
      'portal',
      'account',
    ]),
    'azp': 'token-exchange',
    'email': 'email@example.com',
    'email_verified': True,
    'exp': 1726141246,
    'family_name': 'ACME Inc.',
    'iat': 1726139446,
    'iss': 'http://ebau-keycloak.local/auth/realms/ebau',
    'jti': '458bba12-48a3-4d4b-be23-16ca8e742abe',
    'name': 'ACME Inc.',
    'preferred_username': 'egov:123',
    'realm_access': dict({
      'roles': list([
        'offline_access',
        'default-roles-ebau',
        'uma_authorization',
      ]),
    }),
    'resource_access': dict({
      'account': dict({
        'roles': list([
          'manage-account',
          'manage-account-links',
          'view-profile',
        ]),
      }),
    }),
    'scope': 'openid email profile',
    'sid': 'e0b7c406-d60f-411b-b1b3-3a096a4c5ef2',
    'sub': 'egov:123',
    'typ': 'Bearer',
  })
# ---
